### ISSUE TITLE
Dynamo Table Prefixing

### ISSUE THREAD

**Issue by dmjio** [aristidb/aws#168]
I'd like to append a prefix to my DynamoDB tables (apparently this is recommended by AWS), in order to create a dev. environment. In order to accomplish this I've created a typeclass called `HasPrefix`, and read the prefix from a config file when the app starts. Wanted to pose this as a question to see if others ran into something similar. We could use lenses for this, but regardless, I think this functionality should be added. Would like to come to a consensus w/ others before creating a pull request.

I think exposing a single function that appends the prefix to all records `GetItem`, `PutItem`, etc. wouldn't be a bad idea.

```haskell
------------------------------------------------------------------------------
-- | Execute Dynamo action
dyna ::
  (Transaction r b, HasPrefix b,
   DefaultServiceConfiguration (ServiceConfiguration r NormalQuery),
   MonadIO m, MonadReader Config m) =>
  r -> m b
dyna req = do
  AWSConfig {..} <- asks awsCfg
  let action = setPrefix awsDynamoPrefix req
  liftIO $ runResourceT $ pureAws awsConfig defServiceConfig awsManager action
```

Adding @ozataman

```haskell
------------------------------------------------------------------------------
-- | Table Prefixer
newtype TablePrefix = TablePrefix Text
  deriving (Show, Read, Var)

------------------------------------------------------------------------------
-- | Table Prefixer
class HasPrefix a where
  setPrefix :: TablePrefix -> a -> a

------------------------------------------------------------------------------
-- | Set Dynamo Prefix GetItem
instance HasPrefix GetItem where
  setPrefix (TablePrefix pref) gi@GetItem{..} =
    gi { giTableName = pref <> "-" <> giTableName }

------------------------------------------------------------------------------
-- | Set Dynamo Prefix Query
instance HasPrefix Query where
  setPrefix (TablePrefix pref) qi@Query{..} =
    qi { qTableName = pref <> "-" <> qTableName }

------------------------------------------------------------------------------
-- | Set Dynamo Prefix PutItem
instance HasPrefix PutItem where
  setPrefix (TablePrefix pref) pui@PutItem{..} =
    pui { piTable = pref <> "-" <> piTable }

------------------------------------------------------------------------------
-- | Set Dynamo Prefix DeleteItem
instance HasPrefix DeleteItem where
  setPrefix (TablePrefix pref) di@DeleteItem{..} =
    di { diTable = pref <> "-" <> diTable }

------------------------------------------------------------------------------
-- | Set Dynamo Prefix DeleteItem
instance HasPrefix UpdateItem where
  setPrefix (TablePrefix pref) ui@UpdateItem{..} =
    ui { uiTable = pref <> "-" <> uiTable }

------------------------------------------------------------------------------
-- | Set Dynamo Prefix DeleteTable
instance HasPrefix DeleteTable where
  setPrefix (TablePrefix pref) dt@DeleteTable{..} =
    dt { deleteTableName = pref <> "-" <> deleteTableName }

------------------------------------------------------------------------------
-- | Set Dynamo Prefix CreateTable
instance HasPrefix CreateTable where
  setPrefix (TablePrefix pref) dt@CreateTable{..} =
    dt { createTableName = pref <> "-" <> createTableName }

------------------------------------------------------------------------------
-- | Set Dynamo Prefix UpdateTable
instance HasPrefix UpdateTable where
  setPrefix (TablePrefix pref) dt@UpdateTable{..} =
    dt { updateTableName = pref <> "-" <> updateTableName }

------------------------------------------------------------------------------
-- | Set Dynamo Prefix DescribeTable
instance HasPrefix DescribeTable where
  setPrefix (TablePrefix pref) dt@DescribeTable{..} =
    dt { dTableName = pref <> "-" <> dTableName }

------------------------------------------------------------------------------
-- | Set Dynamo Prefix Lilst
instance HasPrefix ListTables where
  setPrefix = const id
```

**Comment by aristidb** [comment]
If this is a practice that Amazon generally recommends, I guess we could add it, but I'd like to leave the decision to @ozataman 

Not sure about the instance HasPrefix ListTables, and the class name "HasPrefix". Also perhaps, setPrefix could be made idempotent.

**Comment by ozataman** [comment]
Modifying (appending, prepending, etc.) table names is definitely a practical reality when working with DynamoDb - we do something very similar. However, here's how we chose (for whatever reason) to do it (I'm simplifying/inlining the code a bit, but overall structure is the same):

data OurDynTables = DynTable1 | DynTable2 | ...

getDynTableName = do
   cfg <- getOurMonadicConfig
   return $ case cfg of
      DynTable1 -> setPrefix (env cfg) "table_1"

someDynAction = do
    tbl <- getDynTableName DynTable1

    -- issue the request as usual
    dynCall $ ...
    

I think my general thought at the time was that table naming conventions are a matter of **policy**, not base functionality, falling outside of the direct and/or low-level functionality of matching the direct AWS API. Different users may want to do it differently. For example, some use cases may have even greater dynamic table naming requirements, where prepending is not enough - perhaps they need to prepend an environment, but then append a tenant ID and a time-bucket because they keep adding and dropping tables over time. It seems silly to use the typeclass for prepending, but still modify the name manually to further tweak it.

However, I think this could maybe nicely belong in a "utils" package like the one we have been working on[1].  We haven't popularized or put it on Hackage yet, but it's meant to contain some higher level calls we produced over time working with DynamoDb.

If we want to put it in the base aws package, I will readily concede that it would help with the typical/base scenarios. But because aws is meant to be the core, rarely-changing base package that leaves policy decisions out, I'm a bit hesitant. 

[1] https://github.com/Soostone/hs-dynamodb-tools
